### ISSUE TITLE
has_one :through doesn't work with a self join on an STI model

### ISSUE THREAD

**Issue by haines** [rails/rails#12114]
Using ActiveRecord 4.0.0, I'm trying to add an association using a `has_one` through a self join on an STI model, like this ([gist here](https://gist.github.com/haines/6421402)):

``` ruby
class Person < ActiveRecord::Base
end

class Child < Person
  belongs_to :parent
  has_many :toys
end

class Parent < Person
  has_many :children
  has_many :toys, through: :children
end

class Toy < ActiveRecord::Base
  belongs_to :child
  has_one :parent, through: :child
end
```

I can go down through the `has_many` quite happily - `parent.toys` correctly returns a collection of toys. However, going back up the other way doesn't work - `toy.parent` returns nil.

The latter generates the following incorrect SQL:

``` sql
SELECT "people".* FROM "people"
INNER JOIN "people" "children_parent_join"
  ON "people"."id" = "children_parent_join"."parent_id"
WHERE "people"."type" IN ('Child') -- wrong
  AND "people"."type" IN ('Parent')
  AND "children_parent_join"."id" = ?
ORDER BY "people"."id" ASC
LIMIT 1
[["id", 4]]
```

I think to get the correct query it needs to remove the first condition, `"people"."type" IN ('Child')`.

Edit: actually, maybe the right query would be to change the first condition to `"children_parent_join"."type" IN ('Child')` rather than removing it altogether.

**Comment by subimage** [comment]
This is still happening in 4.0.13

```ruby
class Article < ActiveRecord::Base
  belongs_to :storyline, counter_cache: true
  has_one :content, 
    :class_name => 'ArticleContent',
    :foreign_key => 'article_id',
    :dependent => :delete
  after_initialize :init_content
  before_create :init_content
  # omitted many lines for brevity
  def init_content
    self.build_content if self.content.nil?
  end
end

class Storyline < Article
  has_many :articles
end
```

```
>> s = Storyline.new
ActiveRecord::AssociationNotFoundError: Association named 'content' was not found on Storyline; perhaps you misspelled it?
  from /Users/seth/Work/mnhq/site/vendor/ruby/2.1.0/gems/activerecord-4.0.13/lib/active_record/associations.rb:162:in `association'
  from /Users/seth/Work/mnhq/site/vendor/ruby/2.1.0/gems/activerecord-4.0.13/lib/active_record/associations/builder/association.rb:70:in `content'
  from /Users/seth/Work/mnhq/site/app/models/article.rb:905:in `init_content'
  from /Users/seth/Work/mnhq/site/vendor/ruby/2.1.0/gems/activesupport-4.0.13/lib/active_support/callbacks.rb:375:in `_run__184785906289624918__initialize__callbacks'
  from /Users/seth/Work/mnhq/site/vendor/ruby/2.1.0/gems/activesupport-4.0.13/lib/active_support/callbacks.rb:80:in `run_callbacks'
  from /Users/seth/Work/mnhq/site/vendor/ruby/2.1.0/gems/activerecord-4.0.13/lib/active_record/core.rb:188:in `initialize'
  from /Users/seth/Work/mnhq/site/vendor/ruby/2.1.0/gems/actionpack-4.0.13/lib/action_dispatch/routing/url_for.rb:104:in `initialize'
  from /Users/seth/Work/mnhq/site/vendor/ruby/2.1.0/gems/activerecord-4.0.13/lib/active_record/inheritance.rb:30:in `new'
  from /Users/seth/Work/mnhq/site/vendor/ruby/2.1.0/gems/activerecord-4.0.13/lib/active_record/inheritance.rb:30:in `new'
  from (irb):1
  from /Users/seth/Work/mnhq/site/vendor/ruby/2.1.0/gems/railties-4.0.13/lib/rails/commands/console.rb:90:in `start'
  from /Users/seth/Work/mnhq/site/vendor/ruby/2.1.0/gems/railties-4.0.13/lib/rails/commands/console.rb:9:in `start'
  from /Users/seth/Work/mnhq/site/vendor/ruby/2.1.0/gems/railties-4.0.13/lib/rails/commands.rb:62:in `<top (required)>'
  from script/rails:6:in `require'
  from script/rails:6:in `<main>'
```

**Comment by alexcameron89** [comment]
Using the script provided, this issue is fixed on master and Rails 5. Using `git bisect` the issue was fixed when Arel 7.0.0.alpha was introduced, 2223f49.
